kong/dao/migrations/helpers.lua

```diff
@@ -100,6 +100,219 @@ function _M.plugin_config_iterator(dao, plugin_name)
+    end
+
+    local column_declarations = {}
+    local len = 0
+    for name, typ in pairs(table_def.columns) do
+      len = len + 1
+      column_declarations[len] = fmt("%s %s", name, typ)
+    end
+    local column_declarations_cql = table_concat(column_declarations, ", ")
+
+    local cql = fmt("CREATE TABLE IF NOT EXISTS %s(%s%s);",
+                    table_def.name,
+                    column_declarations_cql,
+                    primary_key_cql)
+    return coordinator:execute(cql, {}, CASSANDRA_EXECUTE_OPTS)
+  end
+
+
+  local function drop_table_if_exists(coordinator, table_name)
+    local cql = fmt("DROP TABLE IF EXISTS %s;", table_name)
+    return coordinator:execute(cql, {}, CASSANDRA_EXECUTE_OPTS)
+  end
+
+
+  local function get_columns_to_copy(table_structure)
+    local res = {}
+    for k, _ in pairs(table_structure.columns) do
+      res[k] = k
+    end
+    return res
+  end
+
+
+  local function create_aux_table_def(table_def)
+    local aux_table_def = utils.deep_copy(table_def)
+    aux_table_def.name = "copy_of_" .. table_def.name
+    aux_table_def.columns.partition = "text"
+    table.insert(aux_table_def.partition_keys, 1, "partition")
+    return aux_table_def
+  end
+
+
+  --[[
+    Add a new partition key called "partition" to the table specified by table_def.
+
+    table_def has the following structure:
+    { name    = "ssl_certificates",
+      columns = {
+        id         = "uuid",
+        cert       = "text",
+        key        = "text",
+        created_at = "timestamp",
+      },
+      partition_keys = { "id" },
+    }
+  --]]
+
+  function _M.cassandra.add_partition(dao, table_def)
+
+    local copy_records = _M.cassandra.copy_records
+    local coordinator, err = dao.db:get_coordinator()
+    if not coordinator then
+      return nil, err
+    end
+
+    table_def = utils.deep_copy(table_def)
+
+    local aux_table_def = create_aux_table_def(table_def)
+    local columns_to_copy = get_columns_to_copy(table_def)
+    columns_to_copy.partition = function() return cassandra.text(table_def.name) end
+
+    local _, err = create_table_if_not_exists(coordinator, aux_table_def)
+    if err then
+      return nil, err
+    end
+
+    local _, err = copy_records(dao, table_def, aux_table_def, columns_to_copy)
+    if err then
+      return nil, err
+    end
+
+    local _, err = drop_table_if_exists(coordinator, table_def.name)
+    if err then
+      return nil, err
+    end
+
+    table_def.columns.partition = "text"
+    table.insert(table_def.partition_keys, 1, "partition")
+
+    local _, err = create_table_if_not_exists(coordinator, table_def)
+    if err then
+      return nil, err
+    end
+
+    local _, err = copy_records(dao, aux_table_def, table_def, columns_to_copy)
+    if err then
+      return nil, err
+    end
+
+    local _, err = drop_table_if_exists(coordinator, aux_table_def.name)
+    if err then
+      return nil, err
+    end
+  end
+
+end
+
```

Comment:
This entire file has duplicate contents (open it locally on this branch) this absolutely needs to be fixed prior to the merge.